Mathlib/NumberTheory/SelbergSieve.lean

```diff
@@ -0,0 +1,237 @@
+ * `𝒜 d` for `multSum d`
+ * `P` for `prodPrimes`
+ * `a` for `weights`
+ * `X` for `totalMass`
+ * `ν` for `nu`
+ * `y` for `level`
+ * `R d` for `rem d`
+ * `g d` for `selbergTerms d`
+
+## References
+
+ * [Heath-Brown, *Lectures on sieves*][heathbrown2002lecturessieves]
+ * [Koukoulopoulos, *The Distribution of Prime Numbers*][MR3971232]
+
+-/
+
+noncomputable section
+
+open scoped BigOperators ArithmeticFunction
+
+open Finset Real Nat
+
+/--
+We set up the Selberg sieve as follows. Take a finite set of natural numbers `A`, whose elements
+are weighted by a sequence `a n`. Also take a finite set of primes `P`, represented as a squarefree
+natural number. These are the primes that we will sift from our set `A`. Suppose we can approximate
+`∑ n ∈ {k ∈ A | d ∣ k}, a n = ν d * X + R d`, where `X` is an approximation to the total size of `A`
+and `ν` is a multiplicative arithmetic function such that `0 < ν p < 1` for all primes `p ∣ P`.
+
+Then the fundamental theorem of the Selberg sieve will give us an upper bound on the size of the
+sifted sum `∑ n ∈ {k ∈ support | k.Coprime P}, a n`, obtained by removing any elements of `A` that
+are a multiple of a prime in `P`.
+-/
+class SelbergSieve where
+  /-- The set of natural numbers that is to be sifted. The fundamental lemma yields an upper bound
+    on the size of this set after the multiples of small primes have been removed. -/
+  support : Finset ℕ
+  /-- The finite set of prime numbers whose multiples are to be sifted from `support`. -/
+  prodPrimes : ℕ
+  prodPrimes_squarefree : Squarefree prodPrimes
+  /-- A sequence representing how much each element of `support` should be weighted. -/
+  weights : ℕ → ℝ
+  weights_nonneg : ∀ n : ℕ, 0 ≤ weights n
+  /-- An approximation to `∑ i in support, weights i`, i.e. the size of the unsifted set. A bad
+    approximation will yield a weak statement in the final theorem. -/
+  totalMass : ℝ
+  /-- `nu d` is an approximation to the proportion of elements of `support` that are a multiple of
+    `d` -/
+  nu : ArithmeticFunction ℝ
+  nu_mult : nu.IsMultiplicative
+  nu_pos_of_prime : ∀ p : ℕ, p.Prime → p ∣ prodPrimes → 0 < nu p
+  nu_lt_one_of_prime : ∀ p : ℕ, p.Prime → p ∣ prodPrimes → nu p < 1
+  /-- The `level` of the sieve controls how many terms we include in the inclusion-exclusion type
+    sum. A higher level will yield a tighter bound for the main term, but will also increase the
+    size of the error term. -/
+  level : ℝ
+  one_le_level : 1 ≤ level
+
+attribute [arith_mult] SelbergSieve.nu_mult
+
+namespace SelbergSieve
+
+namespace Notation
+
+@[inherit_doc nu]
+scoped notation3 "ν" => nu
+@[inherit_doc prodPrimes]
+scoped notation3 "P" => prodPrimes
+@[inherit_doc weights]
+scoped notation3 "a" => weights
+@[inherit_doc totalMass]
+scoped notation3 "X" => totalMass
+@[inherit_doc support]
+scoped notation3 "A" => support
+@[inherit_doc level]
+scoped notation3 "y" => level
+/-- `1 ≤ y`-/
+scoped notation3 "hy" => one_le_level
```

Comment:
My personal preference is to name hypotheses descriptively (like we name theorems), so I would call this `y_ge_one` instead of `hy` (so you don't have to remember what hypothesis on `y` is `hy`...). But it's of course up to you...